BBC: Blead Breaks Regexp::Util

This is a bug report for perl from "Carlos Guevara" <carlos@carlosguevara.com>,
generated with the help of perlbug 1.43 running under perl 5.43.8.


-----------------------------------------------------------------
<!--[Please describe your issue here]-->

BBC: Blead Breaks Regexp::Util

Please see https://fast2-matrix.cpantesters.org/?dist=Regexp::Util

<!--[Please do not change anything below this line]-->
<!------------------------------------------------------------------- -->


---
**Flags**
- category=core
- severity=low
---
**Perl configuration**
```
Site configuration information for perl 5.43.8:

Configured by cpan at Sun Feb  1 00:56:32 EST 2026.

Summary of my perl5 (revision 5 version 43 subversion 8) configuration:
  Commit id: 81d1fccb0bb22a976e185a4af70a0ec2604e54a2
  Platform:
    osname=linux
    osvers=6.12.0-124.29.1.el10_1.x86_64
    archname=x86_64-linux
    uname='linux cjg-rhel10 6.12.0-124.29.1.el10_1.x86_64 #1 smp preempt_dynamic mon jan 12 19:32:05 est 2026 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/cpan/bin/perl -Dscriptdir=/home/cpan/bin/perl/bin -Dusedevel -Duse64bitall'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2'
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='14.3.1 20250617 (Red Hat 14.3.1-2)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /usr/lib64 /usr/local/lib64
    libs=-lpthread -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/lib/../lib64/libc.so.6
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.39'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'


---
@INC for perl 5.43.8:
    /home/cpan/bin/perl/lib/site_perl/5.43.8/x86_64-linux
    /home/cpan/bin/perl/lib/site_perl/5.43.8
    /home/cpan/bin/perl/lib/5.43.8/x86_64-linux
    /home/cpan/bin/perl/lib/5.43.8

---
Environment for perl 5.43.8:
    HOME=/home/cpan
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/cpan/bin/perl/bin:/home/cpan/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash
```

Error message complains about `RXf_PMf_SHIFT_NEXT_` being undefined:
```
In file included from /home/cpan/bin/perl/lib/5.43.8/x86_64-linux/CORE/perl.h:4550,
                 from Util.xs:2:
Util.xs: In function 'XS_Regexp__Util_regexp_seen_evals':
/home/cpan/bin/perl/lib/5.43.8/x86_64-linux/CORE/regexp.h:518:27: error: 'RXf_PMf_SHIFT_NEXT_' undeclared (first use in this function)
  518 | #  define RXf_BASE_SHIFT (RXf_PMf_SHIFT_NEXT_ + 2)
      |                           ^~~~~~~~~~~~~~~~~~~
/home/cpan/bin/perl/lib/5.43.8/x86_64-linux/CORE/regexp.h:522:39: note: in expansion of macro 'RXf_BASE_SHIFT'
  522 | #  define RXf_EVAL_SEEN         (1U<<(RXf_BASE_SHIFT+3))
      |                                       ^~~~~~~~~~~~~~
Util.xs:55:36: note: in expansion of macro 'RXf_EVAL_SEEN'
   55 |         RETVAL = RX_EXTFLAGS(re) & RXf_EVAL_SEEN;
      |                                    ^~~~~~~~~~~~~
```

Likely 3f18151623c65345fb59e8024f98c3e1caef1a44 if I had to guess.

--- a/embed.h
+++ b/embed.h
@@ -45,8 +45,6 @@
 #     undef OPpPARAM_IF_FALSE
 #     undef OPpPARAM_IF_UNDEF
 #     undef OPpSELF_IN_PAD
-#     undef RXf_PMf_CHARSET_SHIFT_
-#     undef RXf_PMf_SHIFT_NEXT_
 #     undef utf16_to_utf8
 #     undef utf16_to_utf8_reversed
 #   endif /* !defined(PERL_EXT) */
@@ -58,9 +56,8 @@
 #     undef PARSE_IDENT_ERROR_POSITION
 #     undef PARSE_IDENT_ERROR_TEXT
 #     undef RExC_parse_advance
-#     undef RXf_PMf_SHIFT_COMPILETIME_
 #     undef WARN_HELPER_
-#   endif /* !defined(PERL_EXT_RE_BUILD) */
+#   endif
 # endif /* !defined(PERL_CORE) */
 #else /* if !defined(PERL_DO_UNDEFS) */
 

--- a/regen/embed.pl
+++ b/regen/embed.pl
@@ -3610,7 +3610,6 @@ my @needed_by_ext_re = qw(
     PARSE_IDENT_ERROR_POSITION
     PARSE_IDENT_ERROR_TEXT
     RExC_parse_advance
-    RXf_PMf_SHIFT_COMPILETIME_
     WARN_HELPER_
 );
 
@@ -3620,8 +3619,6 @@ my @needed_by_ext = qw(
     OPpPARAM_IF_FALSE
     OPpPARAM_IF_UNDEF
     OPpSELF_IN_PAD
-    RXf_PMf_CHARSET_SHIFT_
-    RXf_PMf_SHIFT_NEXT_
 );
 
 # This is a list of symbols that are needed to be visible everywhere and are
@@ -3647,6 +3644,9 @@ my @undocumented_always_visible = qw(
     MEM_WRAP_NEEDS_RUNTIME_CHECK_
     MEM_WRAP_WILL_WRAP_
     NV_BODYLESS_UNION_
+    RXf_PMf_CHARSET_SHIFT_
+    RXf_PMf_SHIFT_COMPILETIME_
+    RXf_PMf_SHIFT_NEXT_
     shifted_octet_
     STATIC_ASSERT_STRUCT_BODY_
     STATIC_ASSERT_STRUCT_NAME_

--- a/regen/embed.pl
+++ b/regen/embed.pl
@@ -160,7 +160,9 @@ my %per_file_definitions = (
 # The list does not include symbols that we have documented as being reserved
 # for perl's use, namely those that match the pattern just above.
 # There are two parts of the list; the second part contains the symbols which
-# have a trailing underscore; the first part those without.
+# have a trailing underscore indicating the intent for this symbol to not be
+# directly usable by XS code.  The first part are those symbols without a
+# trailing underscore.
 #
 # For all modules that aren't deliberately using particular names, all the
 # other symbols on it are namespace pollutants.
@@ -3633,10 +3635,15 @@ my @needed_by_ext = qw(
 #
 # Typically these are symbols that are behind-the-scenes helpers whose use is
 # obvious from inspection of the things they help.
+#
+# The list has two parts, separated by a blank line.  The names in the second
+# part have a trailing underscore, indicating the intent for this symbol to
+# not be directly usable by XS code
 my @undocumented_always_visible = qw(
+    MAX_UNICODE_UTF8_BYTES
+
     EXTEND_NEEDS_GROW_
     EXTEND_SAFE_N_
-    MAX_UNICODE_UTF8_BYTES
     MEM_WRAP_NEEDS_RUNTIME_CHECK_
     MEM_WRAP_WILL_WRAP_
     NV_BODYLESS_UNION_

